src/paths/steiner.cpp

```diff
@@ -0,0 +1,586 @@
+    IGRAPH_FINALLY(igraph_matrix_destroy, &distance);
+
+    /*
+     * Compute distances between all pairs of vertices. The Dreyfus - Wagner algorithm needs complete graph information
+     * hence this step is necessary.
+     */
+    IGRAPH_CHECK(igraph_distances_dijkstra(graph, &distance, igraph_vss_all(), igraph_vss_all(), pweights, IGRAPH_ALL));
+
+    IGRAPH_CHECK(igraph_vector_int_init_copy(&steiner_terminals_copy, terminals));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &steiner_terminals_copy);
+    igraph_vector_int_sort(&steiner_terminals_copy);
+    q = VECTOR(steiner_terminals_copy)[0];
+
+    igraph_vector_int_remove(&steiner_terminals_copy, 0);
+
+    /*
+     * DP table with size number of vertices in the graph + 2 ^ (number of steiner_terminals_copy - 1)
+     * 2 ^ (number of steiner_terminals_copy - 1) is number of subsets.
+     */
+    IGRAPH_CHECK(igraph_matrix_init(&dp_cache, no_of_nodes + pow(2, igraph_vector_int_size(&steiner_terminals_copy) - 1), no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_destroy, &dp_cache);
+
+    igraph_matrix_fill(&dp_cache, IGRAPH_INFINITY);
+    /*
+     * for singleton value the distance in dp cahce is just the
+     * distance between same vertices in distance matrix
+     */
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        IGRAPH_ALLOW_INTERRUPTION();
+        for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+            MATRIX(dp_cache, i, j) = MATRIX(distance, i, j);
+        }
+    }
+
+    /*
+     * Usage: Data structure to store index value of subsets.
+     * The index would be used in DP table.
+     */
+    dictionary subsetMap;
+    IGRAPH_CHECK(generateSubsets(&steiner_terminals_copy, igraph_vector_int_size(&steiner_terminals_copy), no_of_nodes, subsetMap, allSubsets));
+
+    int steiner_terminals_copy_size = igraph_vector_int_size(&steiner_terminals_copy);
+    for (igraph_integer_t m = 2; m <= steiner_terminals_copy_size; m++) {
+        for (igraph_integer_t i = 0; i < (igraph_integer_t)allSubsets.size(); i++) {
+            auto it = allSubsets.begin();
+            std::advance(it, i);
+            int_set D = *it;
+            igraph_integer_t indexOfSubsetD;
+            indexOfSubsetD = fetchIndexofMapofSets(D, subsetMap);
+
+            IGRAPH_ALLOW_INTERRUPTION();
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                MATRIX(dp_cache, indexOfSubsetD, j) = IGRAPH_INFINITY;
+            }
+
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                igraph_real_t distance1 = IGRAPH_INFINITY;
+
+                for (auto E :  D) {
+
+                    if (E != j) {
+                        igraph_real_t distanceEJ = MATRIX(distance, E, j);
+
+                        int_set DMinusE = D;
+
+                        /*
+                         *  A set with Singleton value E removed from subset D
+                         */
+                        for (auto item :  DMinusE) {
```

Comment:
Why don't you just do
```                       
DMinusE.erase(E);
```
instead of this loop?

Reply:
Thanks. Yes it makes sense instead of looping. I have made the change